### ISSUE TITLE
Arguments aren't passed to run when using just 'danger'

### ISSUE THREAD

**Issue by orta** [danger/danger-js#177]
https://circleci.com/gh/danger/danger-js/324
https://circleci.com/gh/danger/danger-js/325

After doing 

```js
- yarn
- yarn add danger
```
 I would have expected

`yarn danger -- run --dangerfile dangerfile.circle.js` or `yarn danger -- -d dangerfile.circle.js`  to work. As that's what's in the code.

```js
program
  .option("-v, --verbose", "Verbose output of files")
  .option("-c, --external-ci-provider [modulePath]", "Specify custom CI provider")
  .option("-d, --dangerfile [filePath]", "Specify custom dangerfile other than default dangerfile.js")
  .parse(process.argv)
```

but it didn't work:

```
yarn danger v0.18.1
$ "/home/ubuntu/danger-js/node_modules/.bin/danger" -d dangerfile.circle.js

  error: unknown option `-d'
```

weird

**Comment by alex3165** [comment]
When I install the latest version of danger, the executable JS file look like this:
```
program
    .version(package_json_1.version)
    .command("run", "Runs danger on your local system", { isDefault: true })
    .command("pr", "Runs your changes against an existing PR")
    .parse(process.argv);
```

There is nothing defined for a `-d` option in  `node_modules/danger/distribution/commands/danger.js`

**Comment by orta** [comment]
Yeah, the debugging options have always been an ENV var

**Comment by macklinu** [comment]
It is strange that the TypeScript source shows the `-d` option in `source/commands/danger.ts`, but that line is not present in `distribution/commands/danger.js`, as stated in https://github.com/danger/danger-js/issues/177#issuecomment-287784035. I'm going to take a look at this and see if I can resolve this.

**Comment by macklinu** [comment]
Oops, I was looking at the wrong file. 😛  Still going to take a look.

**Comment by macklinu** [comment]
I wonder if this is related to https://github.com/tj/commander.js/issues/469?

**Comment by paulmelnikow** [comment]
I just ran into this setting up Danger for [Shields](https://github.com/badges/shields).

Would you be up for switching to [argparse](https://github.com/nodeca/argparse)? It's a conscientious port of [Python's absolutely delightful argument parser](https://docs.python.org/3/library/argparse.html). It has a nice API, provides excellent error messages, and most importantly, does not allow errors to pass silently. I've used it in Node projects here and there and had good luck with it.

If not, no worries. I'm just not that invested in fixing commander.

**Comment by orta** [comment]
Love your work <3. 

Yeah, I'm not committed to commander at all, I've not really figured out what's going on here ( as I consider our use case so vanilla too )

**Comment by paulmelnikow** [comment]
Thanks! My focus since joining Shields is scaling out contributions and I'm excited to get Danger into the mix.

I'll give this a shot when I've a second.

**Comment by paulmelnikow** [comment]
Seems like there are three ways to go about this:

1. Get commander to pass arguments to subcommands. Looks like this has been variously attempted: tj/commander.js#692 tj/commander.js#614 tj/commander.js#612.
2. Stop using git-style subcommands, and instead make a unified runner using argparse. This seems cleanest.
3. Re-implement commander's git-style subcommands, in a not-broken way.

Thoughts?

**Comment by orta** [comment]
2, yeah

**Comment by paulmelnikow** [comment]
Is there a faster way to run the updated CLI than a full `yarn build`? On my machine it takes 7–10 seconds.

I tried `yarn build:watch` and it seems to respond after I save, but when I run `node distribution/commands/danger.js pr hi` my changes to the CLI don't seem to take effect. They do take effect after I run `yarn build`.

**Comment by paulmelnikow** [comment]
2.

I'm not sure I understand [this comment](https://github.com/danger/danger-js/blob/e7ca5f26292c0b19ef2851490b80c11f5bf5df82/source/commands/danger-process.ts#L34-L35). It refers to dynamic properties, which must mean dynamic command-line arguments, though I can't see anywhere that a dynamic property of `app` is used.

**Comment by orta** [comment]
1, I don't think so - I very rarely use the build process, I run during dev time using ts-node: `yarn ts-node source/commands/danger-init.ts` 

2, `program` has no type definition so I made an `app` which does have a type definition, it's not dynamic CLI args, but that I wanted to know what all the possible args are, you can see them here: https://github.com/danger/danger-js/blob/e7ca5f26292c0b19ef2851490b80c11f5bf5df82/source/commands/utils/sharedDangerfileArgs.ts#L9-L21

**Comment by paulmelnikow** [comment]
1, Super helpful. I didn't know you could do that. It's my first time using TypeScript. Thanks!
2, Yea, gotcha, I've been using those. I think not all of them apply to every command.

I have a first pass mostly done. It's a straight refactor, and does not fix the problem of args for plain `danger`.

I'll test it a little bit more and then PR it so you can take a look. Hoping you can help me test it also. I don't think the changes I'm making are covered with tests.

**Comment by orta** [comment]
Yeah, they won't be - I've always struggled with writing tests for the commands and opted to try move as much as possible outside of it. I'll give it a run when you've got something 👍

**Comment by orta** [comment]
I took a stab at the simplest answer to this, which is duplicating the code for running across `danger` and `danger run` in #447

**Comment by orta** [comment]
Didn't really fix it

**Comment by paulmelnikow** [comment]
Hey, sorry I didn't respond earlier. Let me push up the half-done work I have. If it seems in a promising direction I'll pick it back up.

Most of where I'm hung up at this point in the refactor is not understanding clearly which programs can accept which arguments. Some of them don't make sense for some commands, e.g. the command that takes a positional dangerfile shouldn't also accept one via flag. argparse is strict, so it's important to sort this out.

I think going from the working multi-program runner to fix this issue will be straightforward enough. Basically it'd mean detecting which of the two cases we're in, then applying the appropriate argument parser. The "running" of the command is manual (you'll see in the branch) so there's no magic.

**Comment by orta** [comment]
I'm happy to either make those decisions ( positional dangerfiles shouldn't happen etc )  - I'm also happy to have you have a fresh perspective. 

As long as the runtime-y ones  are consistent among: `danger`, `danger run`, `danger pr`, `danger process` I'm happy to make a major version change for a better perspective.

**Comment by paulmelnikow** [comment]
Okay, sure, I could bring my perspective on it.

I'm not totally sure I understand what `danger process` is for. Which is [the DSL being referred to here](https://github.com/danger/danger-js/pull/462/files#diff-8c6e70c35ba4a2d175857bfbd543a316R21)? What's the reason for doing the work in a separate process?

Also, is `danger runner` purely for internal use by danger? Is there a reason that needs to be invoked in a separate process? Also since that program is not user-facing at all, I wonder if it would make sense to split it out into a separate CLI.

**Comment by orta** [comment]
`danger Process`: http://danger.systems/js/usage/danger-process.html

TLDR: Let's other environments use Danger in their language without implementing all the hard stuff. 

`danger runner` totally internal yes, it only exists to have the same infra as `danger process` Danger creates a "JSONDSL" which is passed to `danger runner` which then turns it into the real DSL in the reference. The process separation is important because it allows async code to work like you think, rather [than like this](https://github.com/danger/danger-js/blob/master/docs/guides/peril.html.md#dangerfile-implications)

**Comment by orta** [comment]
So it needs to be able to be invoked via the OS _somehow_, adding an extra command just felt like the simplest answer

**Comment by orta** [comment]
Should be fixed in 3.0 which simplified all the CLI user interface
